src/conv/invokers/mlir_impl_gemm.cpp

```diff
@@ -0,0 +1,134 @@
+// - For NHWC, sizes = {N, H, W, C}; strides = {C*H*W, W*C, C, 1}
+
+// In MIOpen however, size and strides are not aligned. Permutation of the strides are used to
+// infer actual layout
+// - For NCHW, sizes = {N, C, H, W}; strides = {C*H*W, H*W, W, 1}
+// - For NHWC, sizes = {N, C, H, W}; strides = {C*H*W, 1, W*C, C}
+auto permuteDimsStrides(const std::vector<size_t>& dims, const std::vector<size_t>& strides)
+{
+    auto sorted_dims    = dims;
+    auto sorted_strides = strides;
+    auto p              = TensorDescriptor::sort_permutation(strides, std::greater<>{});
+    std::transform(p.begin(), p.end(), sorted_dims.begin(), [&](auto i) { return dims[i]; });
+    std::transform(p.begin(), p.end(), sorted_strides.begin(), [&](auto i) { return strides[i]; });
+    return std::make_tuple(sorted_dims, sorted_strides);
+};
+
+void permuteDimStridesAllDir(const conv::ProblemDescription& conv_problem,
+                             std::vector<size_t>& in_dims,
+                             std::vector<size_t>& in_strides,
+                             std::vector<size_t>& weights_dims,
+                             std::vector<size_t>& weights_strides,
+                             std::vector<size_t>& out_dims,
+                             std::vector<size_t>& out_strides)
+{
+    const TensorDescriptor& in = conv_problem.GetIn();
+    std::make_tuple(in_dims, in_strides) = permuteDimsStrides(in.GetLengths(), in.GetStrides());
+
+    const TensorDescriptor& weights = conv_problem.GetWeights();
+    std::make_tuple(weights_dims, weights_strides) =
+        permuteDimsStrides(weights.GetLengths(), weights.GetStrides());
+
+    const TensorDescriptor& out = conv_problem.GetOut();
+    std::make_tuple(out_dims, out_strides) = permuteDimsStrides(out.GetLengths(), out.GetStrides());
+}
+
+MlirConvArgs MakeMlirConvArgs(ConstData_t in,
+                              const std::vector<size_t>& in_dims,
+                              const std::vector<size_t>& in_strides,
+                              ConstData_t w,
+                              const std::vector<size_t>& weights_dims,
+                              const std::vector<size_t>& weights_strides,
+                              ConstData_t out,
+                              const std::vector<size_t>& out_dims,
+                              const std::vector<size_t>& out_strides)
+{
+    auto cpyToMemRef = [](const void* ptr,
+                          const std::vector<size_t>& dims,
+                          const std::vector<size_t>& strides,
+                          MemRef4DGeneric& target) {
+        target.basePtr = const_cast<void*>(ptr); // NOLINT (cppcoreguidelines-pro-type-const-cast)
+        target.data    = const_cast<void*>(ptr); // NOLINT (cppcoreguidelines-pro-type-const-cast)
+        target.offset  = 0;
+        std::copy(dims.cbegin(), dims.cend(), &target.sizes[0]);
+        std::copy(strides.cbegin(), strides.cend(), &target.strides[0]);
```

Comment:
This can be addressed, but with some more inconvenience here. It will be some more code typing in constructing those const data strictures. (Me personally prefer the less typing, existing approach). This should become more clear in the next commit.
 - From previous comments, I need to define input and filter into const types.
 - From this one, I need to be able to update the pointer fields, so that means:
   - Outside the factory, I construct temporary non-const input and filter
   - Inside the factory, I construct const input and filter, manually

Reply:
@jerryyin Whatever design you are going to select, it should be able to satisfy [this](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/841#discussion_r611974707) and [this](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/841#discussion_r611971489). This is the simplest test of design validity.

Reply:
Let's postpone this. #862 created.